toolchain/sem_ir/formatter.cpp

```diff
@@ -22,70 +22,56 @@
 namespace Carbon::SemIR {
 
 // Formatter for printing textual Semantics IR.
-class Formatter {
+class Formatter::FormatterImpl {
  public:
-  enum class AddSpace : bool { Before, After };
-
-  explicit Formatter(const Lex::TokenizedBuffer& tokenized_buffer,
-                     const Parse::Tree& parse_tree, const File& sem_ir,
-                     llvm::raw_ostream& out)
-      : sem_ir_(sem_ir),
-        out_(out),
-        inst_namer_(tokenized_buffer, parse_tree, sem_ir) {}
+  explicit FormatterImpl(const Lex::TokenizedBuffer& tokenized_buffer,
+                         const Parse::Tree& parse_tree, const File& sem_ir)
+      : sem_ir_(sem_ir), inst_namer_(tokenized_buffer, parse_tree, sem_ir) {}
 
-  // Prints the SemIR.
-  //
-  // Constants are printed first and may be referenced by later sections,
-  // including file-scoped instructions. The file scope may contain entity
-  // declarations which are defined later, such as classes.
-  auto Format() -> void {
-    out_ << "--- " << sem_ir_.filename() << "\n\n";
+  // Prints the full SemIR.
+  auto Print(llvm::raw_ostream& out) -> void {
+    llvm::SaveAndRestore store_out(out_, &out);
+    Format();
+  }
 
-    FormatConstants();
-    FormatImportRefs();
+  // Prints a code block.
+  auto PrintPartialTrailingCodeBlock(llvm::ArrayRef<SemIR::InstId> block,
+                                     int indent, llvm::raw_ostream& out)
+      -> void {
+    llvm::SaveAndRestore store_indent(indent_, indent);
+    llvm::SaveAndRestore store_out(out_, &out);
 
-    out_ << inst_namer_.GetScopeName(InstNamer::ScopeId::File) << " ";
+    *out_ << ' ';
     OpenBrace();
-
-    // TODO: Handle the case where there are multiple top-level instruction
-    // blocks. For example, there may be branching in the initializer of a
-    // global or a type expression.
-    if (auto block_id = sem_ir_.top_inst_block_id(); block_id.is_valid()) {
-      llvm::SaveAndRestore file_scope(scope_, InstNamer::ScopeId::File);
-      FormatCodeBlock(block_id);
+    constexpr int MaxPrinted = 10;
+    if (block.size() > MaxPrinted) {
+      Indent();
+      *out_ << "... skipping " << (block.size() - MaxPrinted) << " insts ...\n";
+      block = block.take_back(MaxPrinted);
```

Comment:
Maybe
```suggestion
      *out_ << "... skipping " << (block.size() - MaxPrinted + 1) << " insts ...\n";
      block = block.take_back(MaxPrinted - 1);
```
so we never print more than `MaxPrinted` lines, including the "skipping" line. (It bugs me when tools say "skipping 1 thing", and could have used the same amount of vertical space to actually print the thing they skipped.)

Reply:
Done, with a slightly different approach: I think it's cleaner to do this with a single edit, in the `if`.